boefjes/docs/normalisers-runner.md

```diff
@@ -0,0 +1,128 @@
+
+## Requirements
+[___By Jan___](https://github.com/minvws/nl-kat-coordination/issues/1136#issuecomment-1584306551)
+
+Initially, our design called for normalisers as AWS Lambda-like functions. This would make it possible to run them in micro-VMs/micro-containers and distribute them as small code packages (e.g., code + requirements) targeting a specific pre-built Python (or other interpreter) container running on, for example, [Firecracker](https://firecracker-microvm.github.io).
+
+This has a few advantages:
+* The code of normalisers runs sandboxed.
+* The input can be a single raw file (easily testable).
+* They can be run in parallel.
+* The output is easily tested by testing the returned objects for value and schema validity. (e.g. using [JSON Schema Validation](https://python-jsonschema.readthedocs.io/en/stable/validate/))
+* The whole normaliser can be hashed, and as such, we can keep track of what we did with which code/input/output.
+* Crashes can be caught at the runtime level and reported without boilerplate inside the normaliser.
+* Support for multiple languages can be added.
+* normalisers can carry conflicting dependencies without issue.
+* Easily packaged (zip, OCI container of which the last might be overkill).
+* Separation of runner code (e.g., Python 3.10 with a set of reasonable modules) and app code (e.g., the main method doing the heavy lifting).
+
+This also has a few requirements:
+* normalisers do not interact with the outside world (already met except for 1 normaliser who contacts Octopoes).
+* normalisers list their requirements (already met).
+* The Input and Output are text or binary blobs. (currently, the output is a Python object holding data mirroring the Octopoes model).
+
+This also has a few drawbacks (some we can minimize):
+* Startup time for a sandboxed normaliser is longer than for a direct method call.
+* Not all functionality envisioned can be captured in a sandboxed normaliser which has no other I/O options than the initial raw file + job meta and the resulting output.
+* Inter-related objects in the output stream are 'harder' to relate to each other than with Python's references. (maybe solvable by using something akin to [JSON Schema references ($ref)](https://json-schema.org/understanding-json-schema/structuring.html#ref))
+* One-shot return of data, as the runner only processes all output once the container has returned.
+
+Options that this gives us:
+* Output can be JSON, and optionally with versioned schemas.
+* Run various separate runner envs (e.g., Python 3.8, Python 3.9, PHP 7, PHP 8), needs requirements to be set in the normaliser manifest.
+* Cache normaliser dependencies.
+
+## Design
+
+### Runtime
+
+| Technology | Security | Startup overhead | Ease of Use | Distribution |
+| --- | --- | --- | --- | --- |
+| MicroVM (Firecracker) | Excellent isolation | High | Hard | VM image |
+| Containers (Docker) | Medium isolation | Medium-high | Moderate | OCI image |
```

Comment:
The isolation depends on how the containers are deployed. With Kata Containers or gVisor you can also have excellent isolation of containers.

I think this is an important fact, because it means that if we create OCI images people with not so high security requirements that want a simple setup can just use docker or default Kubernetes setup with CRI-O or containerd, while people with high security requirements can create a more complex setup with hardware virtualization to use Kata Container or gVisor. 

With kubernetes using Kata Containers or gVisor is also pretty easy, we would only have to provide a way to let user configure which RuntimeClass to use.

Reply:
Good point, we discussed this earlier but this was not written down. I've added a separate paragraph on hardened runtimes.